File: 1905-count-sub-islands/1905-count-sub-islands.py
class Solution:
    def countSubIslands(self, grid1: List[List[int]], grid2: List[List[int]]) -> int:
        ROWS, COLS = len(grid1), len(grid1[0])
        visited = set()
        
        def dfs(r, c):
            if (r < 0 or c < 0 or r == ROWS or c == COLS or 
                grid2[r][c] == 0 or (r, c) in visited):
                return True
            
            visited.add((r, c))
            result = True
            
            if grid1[r][c] == 0:
                result = False
                
            result = dfs(r + 1, c) and result
            result = dfs(r - 1, c) and result
            result = dfs(r, c + 1) and result
            result = dfs(r, c - 1) and result
            
            return result
        
        count = 0
        for r in range(ROWS):
            for c in range(COLS):
                if grid2[r][c] and (r, c) not in visited and dfs(r, c):
                    count += 1
                    
        return count

File: 1905-count-sub-islands/1905-count-sub-islands.cpp
class Solution {
public:
    int countSubIslands(vector<vector<int>>& grid1, vector<vector<int>>& grid2) {
        int ROWS = grid1.size(), COLS = grid1[0].size(), count = 0;
        vector<vector<int>> visited(ROWS, vector<int>(COLS, 0));
        for (int i = 0; i < ROWS; ++i) {
            for (int j = 0; j < COLS; ++j) {
                if (grid2[i][j] && !visited[i][j] && dfs(grid1, grid2, i, j, visited)) {
                    ++count;
                }
            }
        }
        return count;
    }
    
    bool dfs(vector<vector<int>>& grid1, vector<vector<int>>& grid2, int i, int j, vector<vector<int>>& visited) {
        int ROWS = grid1.size(), COLS = grid1[0].size();
        if (i < 0 || j < 0 || i == ROWS || j == COLS || !grid2[i][j] || visited[i][j]) {
            return true;
        }
        visited[i][j] = 1;
        bool result = grid1[i][j];
        result *= dfs(grid1, grid2, i + 1, j, visited);
        result *= dfs(grid1, grid2, i - 1, j, visited);
        result *= dfs(grid1, grid2, i, j + 1, visited);
        result *= dfs(grid1, grid2, i, j - 1, visited);
        return result;
    }
};